### PR TITLE
Add webhook timeout checks

### PR DESCRIPTION
**Pull request by MorrisLaw**
_What_
- This PR adds an `admission controller webhook` check that looks at the `TimeoutSeconds` value to see if it is above the max of 30 seconds.

_Why_
- customers should be notified if they are above the recommended `TimeoutSeconds` max, because this blocks upgrades.

### CODE REVIEW

**Comment by MorrisLaw** [comment]
@varshavaradarajan @adamwg

File: checks/doks/webhook_timeout.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+// Description returns a detailed human-readable description of what this check
+// does.
+func (w *webhookTimeoutCheck) Description() string {
+	return "Check for admission control webhooks that have exceeded a timeout of 30 seconds."
+}
+
+// Run runs this check on a set of Kubernetes objects.
+func (w *webhookTimeoutCheck) Run(objects *kube.Objects) ([]checks.Diagnostic, error) {
+	const apiserverServiceName = "kubernetes"
```
Comment (varshavaradarajan) [inline checks/doks/webhook_timeout.go:48]: Unused variable?

File: checks/doks/webhook_timeout.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+				Kind:     checks.ValidatingWebhookConfiguration,
+				Object:   &config.ObjectMeta,
+				Owners:   config.ObjectMeta.GetOwnerReferences(),
+			}
+			diagnostics = append(diagnostics, d)
+
+			// We don't want to produce diagnostics for multiple webhooks in the
+			// same webhook configuration, so break out of the inner loop if we
+			// get here.
+			break
```
Comment (varshavaradarajan) [inline checks/doks/webhook_timeout.go:70]: I would think that the timeout problem may be more common that incorrect webhook config (what the webhookReplacementCheck identifies). May be add in all webhook configs with the timeout problem? I'm not sure about this. May be @adamwg will know.

File: checks/doks/webhook_timeout.go (hunk 1 above)
Comment (MorrisLaw) [reply checks/doks/webhook_timeout.go:48]: Ahh yes. Looks like it!

File: checks/doks/webhook_timeout.go (hunk 2 above)
Comment (MorrisLaw) [reply checks/doks/webhook_timeout.go]: > May be add in all webhook configs with the timeout problem?

Are you saying that we shouldn't break out of the loop on the first encounter of a incorrectly set `TimeoutSeconds` but instead loop through all of the webhooks and add to diagnostics as we encounter them? If so, I agree, I think that makes more sense especially since it's likely to happen more often than just 1 time.

**Comment by MorrisLaw** [comment]
@varshavaradarajan made those changes 👍

File: checks.md [hunk 3]
```diff
@@ -371,6 +371,75 @@ webhooks:
         operator: "DoesNotExist"
 ```
 
+## Admission Controller Webhook Timeout
+
+- Name: `admission-controller-webhook-timeout`
+- Groups: `basic`
```
Comment (varshavaradarajan) [inline checks.md:377]: Isn't it `doks`?

File: checks.md [hunk 4]
```diff
@@ -371,6 +371,75 @@ webhooks:
...
+
+Admission control webhook timeouts can block upgrades. Specifically, this happens when an admission control webhook:
```
Comment (varshavaradarajan) [inline checks.md:379]: Can you explain why it can block upgrades?

File: checks/doks/admission_controller_webhook_timeout_test.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+		{
+			name: "no webhook configurations",
+			objs: &kube.Objects{
+				MutatingWebhookConfigurations:   &ar.MutatingWebhookConfigurationList{},
+				ValidatingWebhookConfigurations: &ar.ValidatingWebhookConfigurationList{},
+			},
+			expected: nil,
+		},
+		{
+			name: "TimeoutSeconds value is set to 31 seconds",
```
Comment (waynr) [inline checks/doks/admission_controller_webhook_timeout_test.go:59]: should probably add a test case on the 29 and 30 second marks just to explicitly validate the edge cases.

File: checks.md [hunk 6]
```diff
@@ -371,6 +371,75 @@ webhooks:
         operator: "DoesNotExist"
 ```
 
+## Admission Controller Webhook Timeout
+
+- Name: `admission-controller-webhook-timeout`
+- Groups: `doks`
+
+Admission control webhook timeouts can block upgrades when the API call has been ignored or fails (based on the failure policy) due to an incorrectly configured TimeoutSeconds value. Since webhooks inherently add to API latency, we must stay within the recommended range in order for API requests to be successful. Specifically, this happens when an admission control webhook:
+* has a TimeoutSeconds value greater than 30 seconds.
```
[the remark this answers is not in the document]
Comment (MorrisLaw) [reply checks.md:380]: Good point

File: checks/doks/admission_controller_webhook_timeout_test.go (hunk 5 above)
Comment (MorrisLaw) [reply checks/doks/admission_controller_webhook_timeout_test.go:59]: Sure, I can update that!

File: checks/doks/admission_controller_webhook_timeout.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+				Object:   &config.ObjectMeta,
+				Owners:   config.ObjectMeta.GetOwnerReferences(),
+			}
+			diagnostics = append(diagnostics, d)
+		}
+	}
+
+	for _, config := range objects.MutatingWebhookConfigurations.Items {
+		for _, wh := range config.Webhooks {
+			if *wh.TimeoutSeconds >= int32(1) && *wh.TimeoutSeconds <= int32(30) {
```
Comment (adamwg) [inline checks/doks/admission_controller_webhook_timeout.go:69]: 30 seconds (which is the default timeout for webhooks) is problematic, since it exactly matches the apiserver request timeout. We should ensure webhook timeouts are strictly less than 30 seconds.

Comment (timoreimann) [reply checks/doks/admission_controller_webhook_timeout.go]: To complete the picture: 30 seconds are problematic if/when the admission web hook isn't available because the default Kubernetes API timeout "overrides" the web hook's, even when the failure policy is set to `Ignore`. Perhaps we should add this detail?

Comment (MorrisLaw) [reply checks/doks/admission_controller_webhook_timeout.go]: > Perhaps we should add this detail?

@timoreimann , are you saying to add the detail as a comment or maybe in the `checks.md` description?

Comment (timoreimann) [reply checks/doks/admission_controller_webhook_timeout.go]: In the markdown file I'd say so that users can easily understand why this is problematic. (And sorry, I essentially just repeated what Adam said; didn't see his comment.)

Comment (MorrisLaw) [reply checks/doks/admission_controller_webhook_timeout.go]: Ahhh ok np. Thank you for the detail. I'll update `checks.md` too 👍

**Comment by MorrisLaw** [comment]
Closing to try and re-trigger the circle ci tests in a seperate PR.
